Bbbpad videocall

denier-epnak opened the pull request.

denier-epnak committed:
Install plugin BigBlueButtonBN version 2.4.5

mantecajc committed:
Add videocall form to fill meeting name and invite viewers in block

Users can be looked up by name/email and selected through the form_user_selector JS component
and the block_padplusvideocall_search_identity service, which are customized copies
from other Moodle features.

Current user must have the new capability block/padplusvideocall:invitevideocall
to use these features (otherwise block won't display)

denier-epnak committed:
override: search_identity and form_user_selector services for user invitation in videocall

We need to override service required capability so that contributors can look for anyone

denier-epnak commented on theme/padplus/classes/output/core_renderer.php:
```diff
@@ -299,6 +305,56 @@ public function context_header($headerinfo = null, $headinglevel = 1) {
+                $videocall = array(
+                    'videocall' => array(
+                        'buttontype' => 'videocall',
+                        'title' => 'Appel vidéo',
```
extract string

denier-epnak commented on theme/padplus/classes/output/core_renderer.php:
```diff
@@ -299,6 +305,56 @@ public function context_header($headerinfo = null, $headinglevel = 1) {
+                        'buttontype' => 'videocall',
+                        'title' => 'Appel vidéo',
+                        'url' => new moodle_url(get_videocall_create_url($categorycontext->id, [$viewerid])),
+                        'formattedimage' => 'i/bullhorn',
```
see how to override icon ?

denier-epnak commented on blocks/padplusvideocall/lib.php:
```diff
@@ -0,0 +1,121 @@
+    $moderatorname = fullname($moderator);
+
+    // PADPLUS TODO: string i18n.
+    $message->subject = "Appel vidéo de {$moderatorname}";
```
extract strings. See mockups for wording

denier-epnak commented on blocks/padplusvideocall/lib.php:
```diff
@@ -0,0 +1,121 @@
+
+define('BBBPAD_VIDEOCALL_PATH', '/blocks/padplusvideocall/bbbpad_videocall.php');
+
+function build_videocall_full_url($params) {
```
comment

denier-epnak committed:
rework: replace initiliaze_call service by inlined creationjoin action in bbbpad_videocall controller

The Ajax request to initiliaze_call and subsequent window.open in response handler asks user for popup authorization.

Since this can prevent users to access a videocall,
we define a simpler workflow where everything happens in the controller.
This means we can't have any feedback in the origin window,
since we open a new window for the videocall.

denier-epnak commented on blocks/padplusvideocall/bbbpad_videocall.php:
```diff
@@ -0,0 +1,235 @@
+ * @return object
+ */
+function bbbpad_create_meeting_data($meetingid, $meetingname, $viewerpw, $modpw, $logouturl) {
+    // PADPLUS TODO: extract string?
```
extract string

denier-epnak commented on blocks/padplusvideocall/bbbpad_videocall.php:
```diff
@@ -0,0 +1,235 @@
+        'meetingLayout' => 'VIDEO_FOCUS', // Put focus on video by default since it is a video call.
+        'allowStartStopRecording' => false, // Disable video recording.
+
+        // PADPLUS TODO: Following options does not work on BBB test server. Need to test on our own BBB server.
```
check options

denier-epnak commented on theme/padplus/scss/post/components/_dropdown.scss:
```diff
@@ -64,3 +64,100 @@
 }
+
+// Navbar notifications dropdown.
+#nav-notification-popover-container {
```
est-ce qu'il ne faudrait pas extraire dans son propre fichier dropdown_notifications? Ou dans le layout navbar ? @mantecajc

mantecajc replied:
Peut être dans _navbar c'est pas mal. Pour eviter d'avoir les fichiers _notification, _notifications_dropdown et dropdown ?

denier-epnak replied:
oui c'est pas mal

denier-epnak committed:
Enable BigBlueButton videocall through dashboard block and profile pages

Block padplusvideocall is defined to hold most code related to videocalls,
including Javascript and lib functions.
- can be added to dashboard. Block is displayed when user has the
createvideocall capability (see below)

- bbbpad_videocall.php is the videocall controller which handles creating
and joining a BBB meeting (strongly inspired by bbb_viewer.php from the
BBB mod plugin).

Meeting creation happens in two steps.
First service block_padplusvideocall_initialize_videocall is requested
through Ajax to get moderator url and display viewer url for sharing.
Second moderator url is opened in a new window to call bbbpad_videocall.php
and initiates the meeting.
Problem is opening a new window inside the Ajax handler requires the user
to approve the popup.

- Define new capability block/padplusvideocall:createvideocall, which
allows one to create videocalls. This capability is defined in category
context, to allow managers and course creators (i.e. not students) to
have this capability.

- Override context_header renderer to add a videocall button for profile
pages only. Also provide a theme settings to hide this button.

- Provide Moodle notifications (web&email) for videocalls with animation
and autoplay sound

mantecajc commented on blocks/padplusvideocall/classes/event/videocall_created.php:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public function get_description(): string {
+        return "The user with id '$this->userid' created a video call";
```
On peut mettre cette phrase dans le paquet de lang ?

denier-epnak replied:
Alors je ne sais pas trop pourquoi, mais la convention Moodle veut que les descriptions des logs ne soient pas traduites, mais les noms des événements si. Peut-être pour faciliter les échanges sur les forums en cas de soucis mais ça me paraît mince.
https://docs.moodle.org/dev/Translation_FAQ#Why_are_log_descriptions_displayed_in_English.3F

mantecajc commented on blocks/padplusvideocall/classes/event/videocall_joined.php:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public function get_description(): string {
+        return "The user with id '$this->userid' joined a video call";
```
idem

mantecajc commented on blocks/padplusvideocall/classes/event/videocall_left.php:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public function get_description(): string {
+        return "The user with id '$this->userid' left a video call";
```
idem

mantecajc commented on blocks/padplusvideocall/bbbpad_videocall.php:
```diff
@@ -0,0 +1,208 @@
+    if ($message != '') {
+        echo html_writer::tag('p', $message, array('class' => 'alert alert-warning'));
+    }
+    echo "Vous êtes sorti de l'appel vidéo. Vous pouvez continuer vos activités sur la plateforme ou fermez cette fenêtre.";
```
on peut mettre cette phrase dans le dossier lang ?

denier-epnak replied:
Diff obsolète ? Cette string est extraite dans la dernière version https://github.com/e-PSHAD/PAD/pull/28/files#diff-72dd836cf0bda33ca6f1ad59cf8bcbcf2fb283ed0a2fa4607df9922a1e890c6b
